Add Spatial Audio Box (SA3D) support

rzumer opened the pull request:
Based on the RFC: https://github.com/google/spatial-media/blob/master/docs/spatial-audio-rfc.md

I tested the reading via `boxdumper`. The sample has a first-order ambisonic audio track and had spatial audio data injected via the [spatialmedia](https://github.com/google/spatial-media/tree/master/spatialmedia) tool.

```sh
rzumer@gekai:~/dev/l-smash/cli$ ./boxdumper --box ambi.mp4 | grep -A 13 SA3D
                                [SA3D: Spatial Audio Box]
                                    position = 26998660
                                    size = 36
                                    version = 0
                                    head_locked_stereo = 0
                                    ambisonic_type = 0
                                    ambisonic_order = 1
                                    ambisonic_channel_ordering = 0
                                    ambisonic_normalization = 0
                                    num_channels = 4
                                    channel_map[0] = 0
                                    channel_map[1] = 1
                                    channel_map[2] = 2
                                    channel_map[3] = 3
```

I also added Opus support for `muxer`, since it was useful for testing. I did not notice any side effects, but I am not sure if there is anything else needed for a complete implementation of the format.

```sh
rzumer@gekai:~/dev/l-smash/cli$ ./muxer -i video.mp4 -i audio.mp4?ambisonic-order=1 -o test.mp4
MP4 muxing mode
Track 1: H.264 Advanced Video Coding
Track 2: Opus
Muxing completed!
```
```sh
rzumer@gekai:~/dev/spatial-media$ python spatialmedia ~/dev/l-smash/cli/test.mp4 
Processing: /home/rzumer/dev/l-smash/cli/test.mp4
Loaded file...
        Track 0
        Track 1
                Ambisonic Type: periphonic
                Contains Head-Locked Stereo: False
                Ambisonic Order: 1
                Ambisonic Channel Ordering: ACN
                Ambisonic Normalization: SN3D
                Number of Channels: 4
                Channel Map: [0, 1, 2, 3]
```

Note: the only configurable parameter in the muxer is the ambisonic order, and the number of audio channels must be equal to the expected value based on the ambisonic order. For example, if `ambisonic_order=1` is specified, the audio track must have exactly 4 channels. Embedded head-locked stereo is not supported in the muxer, and the channel map cannot be set, so it is assumed to be in order ([0, 1, 2, 3, ...]).

rzumer committed:
Add Spatial Audio Box (SA3D) support

rzumer committed:
Fix spacing errors

rzumer commented on core/print.c:
```diff
@@ -1333,7 +1354,7 @@ static int isom_print_sample_description_extesion( FILE *fp, lsmash_file_t *file
         lsmash_box_type_t type;
         int (*print_func)( FILE *, lsmash_file_t *, isom_box_t *, int );
-    } print_description_extension_table[33] = { { LSMASH_BOX_TYPE_INITIALIZER, NULL } };
+    } print_description_extension_table[64] = { { LSMASH_BOX_TYPE_INITIALIZER, NULL } };
```
I noticed that a few other tables had their size rounded up to the next power of 2, so I just bumped this to 64 so that it doesn't need to be incremented every time, assuming it has a negligible impact on memory usage and performance.

dwbuiten commented on core/read.c:
```diff
@@ -893,6 +893,35 @@ static int isom_read_stbl( lsmash_file_t *file, isom_box_t *box, isom_box_t *par
+    SA3D->ambisonic_order = lsmash_bs_get_be32( bs );
+    SA3D->ambisonic_channel_ordering = lsmash_bs_get_byte( bs );
+    SA3D->ambisonic_normalization = lsmash_bs_get_byte( bs );
+    SA3D->num_channels = lsmash_bs_get_be32( bs );
```
NIT: Rest of file as aligned `=`.

rzumer replied:
Fixed

dwbuiten commented on core/read.c:
```diff
@@ -893,6 +893,35 @@ static int isom_read_stbl( lsmash_file_t *file, isom_box_t *box, isom_box_t *par
+    SA3D->ambisonic_channel_ordering = lsmash_bs_get_byte( bs );
+    SA3D->ambisonic_normalization = lsmash_bs_get_byte( bs );
+    SA3D->num_channels = lsmash_bs_get_be32( bs );
+    for( uint64_t pos = lsmash_bs_count( bs ); pos < box->size && SA3D->channel_map.entry_count < SA3D->num_channels; pos = lsmash_bs_count( bs ) )
```
`while` might be more readable? Though, if this sort of `for` is use elsewhere in the codebase, I suppose this would be preferred.

rzumer replied:
> Though, if this sort of `for` is use elsewhere in the codebase, I suppose this would be preferred.

It is.

dwbuiten commented on codecs/description.c:
```diff
@@ -3018,6 +3018,23 @@ lsmash_summary_t *isom_create_audio_summary_from_description( isom_sample_entry_
+                summary->sap.ambisonic_order = SA3D->ambisonic_order;
+                summary->sap.ambisonic_channel_ordering = (lsmash_SA3D_AmbisonicChannelOrdering)SA3D->ambisonic_channel_ordering;
+                summary->sap.ambisonic_normalization = (lsmash_SA3D_AmbisonicNormalization)SA3D->ambisonic_normalization;
+                summary->sap.num_channels = SA3D->num_channels;
```
NIT: Align `=`.

rzumer replied:
Done

rzumer committed:
Allow muxing Opus for ISO Base Media

rzumer committed:
Add muxer support for injecting basic spatial audio data

rzumer commented:
@dwbuiten I added muxer support via codec-specific data and fixed a bug in the write implementation. With this the implementation should be complete and correct, at least for my simple test case.

I added sample output from `muxer` and Google's `spatialaudio` tool for validation in the PR description.

dwbuiten reviewed (commented):
Code looks OK, but some commits should be squashed, I think:
* `Add Spatial Audio Box (SA3D) support`
* `Expose SA3D data in public audio summary`
* `Store spatial audio information as codec-specific data`

should either be a single commit, or one read, and one right. No stron opinion on which.

rzumer commented:
> should either be a single commit, or one read, and one right. No stron opinion on which.

The first commit already had a partial write implementation anyway, so I squashed them.
